File: apps/android/app/src/main/java/com/mollotov/browser/handlers/BookmarkHandler.kt
package com.mollotov.browser.handlers

import com.mollotov.browser.browser.BookmarkStore
import com.mollotov.browser.network.Router
import com.mollotov.browser.network.errorResponse
import com.mollotov.browser.network.successResponse

class BookmarkHandler(private val ctx: HandlerContext) {
    fun register(router: Router) {
        router.register("bookmarks-list") { list() }
        router.register("bookmarks-add") { add(it) }
        router.register("bookmarks-remove") { remove(it) }
        router.register("bookmarks-clear") { clear() }
    }

    private suspend fun list(): Map<String, Any?> =
        successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))

    private suspend fun add(body: Map<String, Any?>): Map<String, Any?> {
        val url = body["url"] as? String ?: return errorResponse("MISSING_PARAM", "url is required")
        val title = body["title"] as? String ?: url
        BookmarkStore.add(title, url)
        return successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))
    }

    private suspend fun remove(body: Map<String, Any?>): Map<String, Any?> {
        val id = body["id"] as? String ?: return errorResponse("MISSING_PARAM", "id is required")
        BookmarkStore.remove(id)
        return successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))
    }

    private suspend fun clear(): Map<String, Any?> {
        BookmarkStore.clear()
        return successResponse(mapOf("cleared" to true))
    }
}

File: apps/ios/Kelpie/Handlers/BookmarkHandler.swift
import Foundation

/// API handler for bookmark CRUD: list, add, remove, clear.
struct BookmarkHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("bookmarks-list") { _ in await list() }
        router.register("bookmarks-add") { body in await add(body) }
        router.register("bookmarks-remove") { body in await remove(body) }
        router.register("bookmarks-clear") { _ in await clear() }
    }

    @MainActor
    private func list() async -> [String: Any] {
        await savedResponse()
    }

    @MainActor
    private func add(_ body: [String: Any]) async -> [String: Any] {
        guard let url = body["url"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "url is required")
        }
        let title = body["title"] as? String ?? url
        return await savedResponse(BookmarkStore.shared.add(title: title, url: url))
    }

    @MainActor
    private func remove(_ body: [String: Any]) async -> [String: Any] {
        guard let idStr = body["id"] as? String, let id = UUID(uuidString: idStr) else {
            return errorResponse(code: "MISSING_PARAM", message: "id is required")
        }
        return await savedResponse(BookmarkStore.shared.remove(id: id))
    }

    @MainActor
    private func clear() async -> [String: Any] {
        return await savedResponse(BookmarkStore.shared.removeAll(), cleared: true)
    }

    @MainActor
    private func savedResponse(_ operation: Task<Void, Error>? = nil, cleared: Bool = false) async -> [String: Any] {
        do {
            try await operation?.value
            return cleared ? successResponse(["cleared": true]) : successResponse(["bookmarks": BookmarkStore.shared.toJSON()])
        } catch {
            return errorResponse(code: "BOOKMARK_SYNC_FAILED", message: error.localizedDescription)
        }
    }
}
